proposals/0698-qualified-literals.rst

```diff
@@ -0,0 +1,565 @@
+
+One scenario this can come up is when parsing ``Aeson.Array``, which stores JSON values in a ``Vector``.
+
+Scientific
+~~~~~~~~~~
+
+`Scientific <https://hackage.haskell.org/package/scientific-0.3.8.0/docs/Data-Scientific.html#t:Scientific>`_ represents an arbitrary precision number. It has a ``Num`` instance, but ``+`` and ``-`` are unsafe and can cause OOM. Safety-minded developers might desire to wrap with a newtype that provides ``unsafeAdd`` but not ``+``, to prevent call-sites from accidentally blowing up memory.
+
+::
+
+  newtype BigDecimal = BigDecimal Scientific
+
+  unsafeAdd :: BigDecimal -> BigDecimal -> BigDecimal
+  unsafeAdd = coerce (+)
+
+If you want to write ``BigDecimal`` literals (e.g. for tests), you have to use either the ``BigDecimal`` constructor or write a ``big = BigDecimal`` helper, but that's unsafe if accidentally called on a non-literal, as ``Scientific`` throws a runtime error if converting from a repeating decimal.
+
+With ``QualifiedNumbers``, you could write ``Big.123``, which guarantees that ``Big.fromNumeric`` is only called on literals (e.g. you could configure hlint to ban calling ``BigDecimal.fromNumeric`` directly and only be used via ``QualifiedNumbers``).
+
+::
+
+  -- only called on literals, can be used with any numeric literal: naturals, integers, rationals
+  fromNumeric :: Real a => a -> BigDecimal
+  fromNumeric = BigDecimal . realToFrac
+
+Multiple numeric types
+~~~~~~~~~~~~~~~~~~~~~~
+
+Here's an example of an implementation that's generic to all three numeric types:
+
+::
+
+  module StringInc (fromNumeric) where
+
+  fromNumeric :: (Num a, Show a) => a -> String
+  fromNumeric x = "StringInc (" ++ show (x + 1) ++ ")"
+
+* ``StringInc.10`` => ``"StringInc (11)"``
+* ``StringInc.(-10)`` => ``"StringInc (-9)"``
+* ``StringInc.(10.5)`` => ``"StringInc (23 % 2)"``
+
+Here's an example of an implementation that's different for each numeric type:
+
+::
+
+  module MyMod (fromNumeric) where
+
+  class Impl a where
+    fromNumeric :: a -> String
+
+  instance Impl Natural where
+    fromNumeric n = "NAT:" ++ show n
+
+  instance Impl Integer where
+    fromNumeric n = "NEGATIVE:" ++ show n
+
+  instance Impl Rational where
+    fromNumeric n = "FLOAT:" ++ show (realToFrac n :: Double)
+
+* ``MyMod.10`` => ``"NAT:10"``
+* ``MyMod.(-10)`` => ``"NEGATIVE:-10"``
+* ``MyMod.(10.5)`` => ``"FLOAT:10.5"``
+
+Positive-only literals
+~~~~~~~~~~~~~~~~~~~~~~
+
+The following currently fails to compile with ``-Werror`` enabled:
+
+::
+
+  >>> (-1) :: Natural
+
+  error: [GHC-97441] [-Woverflowed-literals, Werror=overflowed-literals]
+    Literal -1 is negative but Natural only supports positive numbers
+
+However, this check is hardcoded in the compiler for specific types, e.g. ``Natural``, ``Int*``, and ``Word*``. There's no way to enforce this at compile time for custom types, e.g.
+
+::
+
+  -- Invariant: positive
+  newtype UserId = UserId Integer
+
+  UserId (-1) -- works
```

Comment:
I don't find this example motivating. Why are we choosing to define this type in terms of `Integer` rather than `Natural`?

If we used `Natural` then `UserId (-1)` would be an error, right?

Reply:
Sure but even if you are forced to use an `Integer`, you can still use `PatternSynonyms` here.